### PR TITLE
docs(research,eval): deepseek-harness lessons + tier-1 eval harness repairs

### PR DESCRIPTION
**Pull request by wildcard**
## Description

Research into [deepseek-ai/deepseek-harness](https://github.com/deepseek-ai/deepseek-harness) ("Everything is a Plugin"), mapped against caro's three harnesses — the `AgentLoop` runtime, the CaroML validator/journal layer, and the evaluation harness — with tiered adoption proposals: **[`docs/research/deepseek-harness-lessons.md`](https://github.com/wildcard/caro/blob/claude/deepseek-harness-improvements-tktop1/docs/research/deepseek-harness-lessons.md)**. Alongside the doc, this PR ships the smallest repairs the research surfaced (Tier-1/P6a).

### Motivation

`Cargo.toml`'s `candidate-ranking` feature says: *"flip to default in v1.4.0 after the eval harness shows ranking wins."* We are at v1.5.0. The flip isn't overdue because ranking lost — the harness couldn't render the verdict: 3 of 4 CI eval matrix legs were silently no-ops, concurrency config was dead, and the one honest leg's regression floor (31.0) was 47 points below measured reality (78.2%). The doc's proposals all route toward making that verdict renderable; the fix commits repair what was unambiguous.

### Changes Made

- **`docs/research/deepseek-harness-lessons.md`** (new): dsh case study; caro mapping; 8-row principle gap table; proposals P1–P8 in three tiers; what NOT to adopt (incl. the inversion: dsh's "no privileged core" must stay inverted — caro's safety floor is deliberately privileged); 24-gap eval inventory (Appendix A); orphan-crate salvage list (Appendix B).
- **`src/evaluation/harness.rs`**: `HarnessConfig::max_concurrency` was declared but never read — `run_all_tests` spawned one unbounded task per (case × backend). Now bounded by a `tokio::sync::Semaphore`; the permit is acquired **before** `tokio::spawn`, so at most `max_concurrency` tasks (and their captured clones) exist at once, and it is held across generate+evaluate. New regression guard: `evaluation::harness::tests::test_max_concurrency_bounds_in_flight_generations`.
- **`.github/workflows/evaluation.yml`**: matrix reduced to `static_matcher` — the only leg that evaluates anything (the runner registers only `StaticMatcher`; `--backend` doesn't filter; `embedded-*` legs exited 2 behind `|| true`). Regression baseline set to the **measured 78.2** (the workflow's existing −5 rule then blocks below 73.2 and warns under 78.2; the previous 31.0 was stale). Comments explain what restores the other legs.
- **`Cargo.toml`**: declare `benches/performance.rs` as `[[bench]] harness = false`; cargo auto-discovery ran it under libtest where `criterion_main!` never executed.
- **`CLAUDE.md`**: nonexistent `cargo run --bin caro-eval` → `cargo test --test evaluation`; version banner 1.4.0 → 1.5.0.

---

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (code restructuring without changing behavior)
- [x] Documentation update (changes to docs, comments, or guides)
- [ ] Performance improvement (makes code faster or more efficient)
- [x] Test coverage (adds or improves tests)
- [x] CI/CD or tooling (changes to build, release, or development tools)

---

### Code Quality

- [x] I have run `cargo fmt --all` (`cargo fmt --check` clean)
- [x] I have run `cargo clippy -- -D warnings` (clean; run with `--tests --benches`, `--no-default-features --features embedded-cpu`)
- [x] I have run `cargo test` (81 lib tests in `evaluation::` pass; full eval harness run below)
- [ ] I have run `cargo audit` (not run locally — no dependency changes; CI `cargo-audit` and `dependency-review` are green on this PR)

### Testing

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have added contract tests for new public APIs (no new public APIs)
- [ ] I have added integration tests for cross-module workflows (n/a)
- [x] I have verified performance requirements are met (bench target now actually executes — see Testing Evidence)

### Documentation

- [ ] I have added rustdoc comments for new public APIs (no new public APIs; semaphore behavior documented on `run_all_tests`)
- [x] I have updated relevant documentation (research doc; CLAUDE.md drift fixes)
- [ ] I have added examples to demonstrate new functionality (n/a)
- [ ] I have updated the CHANGELOG.md (CHANGELOG entries are compiled at release time per `release-version-alignment`)

### TDD Workflow

- [x] I followed the Red-Green-Refactor cycle — the guard test fails without the semaphore (unbounded spawns observe >2 concurrent generations) and passes with it
- [x] I wrote failing tests before implementing the solution
- [ ] I verified tests with `cargo watch -x test` during development (single-shot runs in CI-like environment)

---

## Breaking Changes

None. The semaphore bounds concurrency at the existing default (10); results are unchanged (same pass counts), only in-flight parallelism is capped.

### API Changes

None.

### Migration Guide

n/a

### Deprecation Plan

n/a

---

## Related Issues and Specs

- Related to [PR #1108](https://github.com/wildcard/caro/pull/1108) (candidate-ranking pipeline — the doc's proposals route toward its overdue eval-gated default flip)
- Related to [PR #1245](https://github.com/wildcard/caro/pull/1245) (Fireworks hybrid-harness learnings — the precedent for this research genre)
- Prior art: `thoughts/shared/plans/evaluation-harness-maturity-milestone.md`

---

## Performance Impact

Eval harness runs are now bounded at `max_concurrency` (default 10) in-flight generations instead of unbounded (101 × backends at once). For the static matcher this is latency-neutral (sub-second full run); for future LLM backends it prevents resource exhaustion.

### Benchmarks

`cargo bench --bench performance -- --test`: all criterion groups now execute (`cli_startup`, safety validation single/batch/concurrent/sustained, shell types, pattern matching) — previously this target never ran at all, so there is no "before" to compare.

### Binary Size

Unchanged (no dependency or shipped-code additions; changes are eval/test/CI-side).

### Memory Usage

Bounded rather than unbounded task fan-out in eval runs (permit acquired before spawn).

---

### Before

```
$ cargo test --test evaluation -- --backend embedded-qwen   # what 2 of 4 CI legs ran
Invalid backend: embedded-qwen. Must be one of: static_matcher, mlx, ollama, vllm
(exit 2, masked by `|| true` — leg reports "backend not available", green)
```

### After

```
$ cargo test --test evaluation -- --backend static_matcher
Total Tests: 101 · Passed: 79 (78.2%) · Correctness 26/26 · Safety 25/25 · POSIX 24/25 · MultiBackend 4/25
(the only honest leg, now the only leg — baseline is the measured 78.2 instead of a stale 31.0)
```

---

### Test Output

```
$ cargo test --no-default-features --features embedded-cpu --lib evaluation::
test evaluation::harness::tests::test_max_concurrency_bounds_in_flight_generations ... ok
test result: ok. 81 passed; 0 failed; 0 ignored; 0 measured; 518 filtered out

$ cargo test --no-default-features --features embedded-cpu --test evaluation -- --backend static_matcher
Total Tests: 101 · Passed: 79 (78.2%) · Failed: 22 (21 of 22 are MultiBackend cases, which need ≥2 registered backends)

$ cargo bench --no-default-features --features embedded-cpu --bench performance -- --test
Testing cli_startup ... Success  ·  Testing pattern_matching ... Success  ·  (all groups) Success
```

**CI (green):** [ci.yml run](https://github.com/wildcard/caro/actions/runs/31918850447) · [LLM Evaluation Harness run](https://github.com/wildcard/caro/actions/runs/31918850486) — the `Evaluate static_matcher` job independently measured **79/101 = 78.2%** on ubuntu-latest / default features / Rust 1.97.1, matching the local run on a different feature set and toolchain (static matcher is deterministic).

### Manual Testing

- [ ] Tested on macOS (Apple Silicon / Intel)
- [x] Tested on Linux (Ubuntu-based container; `embedded-cpu` feature set) + CI ubuntu-latest default features
- [ ] Tested on Windows (10 / 11)
- [x] Tested with different backends (static matcher — the backend this PR's CI leg evaluates)
- [x] Tested edge cases and error conditions (guard test pins the concurrency bound; `max_concurrency: 0` clamps to 1)

---

### Technical Decisions

- **Kept `|| true` in the workflow**: the runner exits 1 whenever pass rate < 100% (`tests/evaluation/main.rs:231`), so the real gate is the baseline-compare step; removing `|| true` would turn CI permanently red. Fixing the exit-code contract is gap A23 in the doc, deliberately out of scope here.
- **Matrix reduced rather than "fixed"**: registering real embedded backends in the runner is gap A4 — a feature, not a repair; the honest state today is one leg.
- **Baseline = measured, not "measured minus headroom"** (per Codex review): the workflow already subtracts 5 points for the blocking line; pre-subtracting again would have let a drop to 70% pass silently.
- **Doc proposals are proposals**: nothing beyond the P6a repairs is implemented here. P5 (runtime profiles) is explicitly flagged as user-facing → `validation-discipline` gates before any spec.
- **Review feedback addressed** (Codex ×1, cubic ×4): see the resolved threads — baseline math, floor reproducibility evidence, doc LOC figures re-measured, `BaselineStore` claim narrowed, semaphore acquired before spawn.
- **Devil's Advocate Review**: being posted as a `## Devil's Advocate Review` comment on this PR (Gate-4 culture for AI-drafted proposals, applied voluntarily to internal-tooling research).

### Future Work

Tier-1: P7 CI doc-verification gates (seed cases: README.md:34 / CLAUDE.md:119 "93.1% pass rate" — untraceable to any harness output); P8 DeepSeek pricing/catalog rows if a DeepSeek backend ever lands. Tier-2: P1 generation journal (unlocks the currently-unreachable `sft_export`), P3 validator middleware chain, P2 execution seam, P6b eval consolidation (incl. deleting the never-compiled `tests/evaluation/src/` orphan). Tier-3: P4 wire the candidate pipeline + flip the flag on evidence; P5 profiles.

### Questions for Reviewers

Is setting the regression baseline to the measured 78.2 in this PR acceptable, or should the baseline change ship separately with its own bake time? (It is a one-value revert either way.)

---

## Reviewer Checklist

- [ ] Code follows Rust best practices and project conventions
- [ ] Tests are comprehensive and follow TDD principles
- [ ] Documentation is clear and complete
- [ ] Changes align with project specifications
- [ ] Performance impact is acceptable
- [ ] Breaking changes are justified and documented
- [ ] Security implications have been considered

---

**By submitting this PR, I confirm that:**

- [x] My code follows the style guidelines of this project (see [AGENTS.md](https://github.com/wildcard/caro/blob/main/AGENTS.md))
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [x] I have read and followed the [contributing guidelines](https://github.com/wildcard/caro/blob/main/CONTRIBUTING.md)
- [x] I agree to the [Code of Conduct](https://github.com/wildcard/caro/blob/main/CODE_OF_CONDUCT.md)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01JfmCW2L5d5qkNJ18mDfz6a

### CODE REVIEW

**Comment by wildcard** [comment]
`[agent]`

**Agent:** Claude Code (`claude-opus-5`)

---

**Dispatcher classification: `path:refuse-list`** (automated resurfacing declined; human merge path unaffected).

**Which glob fired:** `.github/workflows/**` — this PR touches `.github/workflows/evaluation.yml`.

**Rationale (from `.claude/automation/config/policy.yaml`):** workflow definitions are set to `auto_dispatch: false, auto_rebase: false`. CI definitions are the mechanism that validates every other change, so an automated agent must not edit or resurface them unattended — a bad workflow edit degrades the signal that would catch it.

The other touched paths do **not** independently refuse: `Cargo.toml` is `deps_only` (satisfied), and `src/evaluation/harness.rs`, `CLAUDE.md`, `docs/research/deepseek-harness-lessons.md` are unlisted.

**Relaxation path:** any of the following unblocks it —
1. Split the `evaluation.yml` change into its own PR and human-review it; the remaining harness + docs changes then classify as `path:scoped` and get dispatched normally.
2. A human review approval on this PR — refuse-list gates *agent* dispatch, not maintainer merge. This PR is `MERGEABLE` with green checks and can be merged by hand today.
3. Amend the policy glob to carve out an allowance (separate PR against `policy.yaml`; this routine never edits policy itself).

No rebase bead was filed for this PR — refuse-list PRs wait on policy or human review, not on coder-loop work.

If the workflow-file constraint should soften for eval-harness CI specifically, reply here and it will be picked up in the next grooming cycle.

File: .github/workflows/evaluation.yml [hunk 1]
```diff
@@ -61,12 +58,14 @@ jobs:
...
+          # measure one before re-adding a matrix leg.
+          # static_matcher measured 78.2% on the 101-case dataset (2026-08-16);
+          # floor set just below so dataset edits don't trip the -5% gate. The
+          # previous 31.0 floor was stale by 47 points — nobody was reading it.
           case "$BACKEND" in
-            static_matcher) BASELINE="31.0" ;;
-            mlx) BASELINE="0.0" ;;  # TBD - not yet measured
-            embedded-smollm) BASELINE="0.0" ;;  # TBD
-            embedded-qwen) BASELINE="0.0" ;;  # TBD
+            static_matcher) BASELINE="75.0" ;;
```
[the remark this answers is not in the document]
Comment (wildcard) [reply .github/workflows/evaluation.yml]: Agreed — pre-subtracting "headroom" double-counted the workflow's own −5 rule. Fixed in `36edc67`: `BASELINE` is now the measured **78.2**, so the existing rule blocks below 73.2 and warns on anything under 78.2. The workflow comment now says exactly that.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (wildcard) [reply .github/workflows/evaluation.yml]: Fair challenge — here is the evidence rather than the assertion.

1. Local run (`--no-default-features --features embedded-cpu`, Rust 1.94.1): **79/101 = 78.2%** — Correctness 26/26, MultiBackend 4/25, POSIX 24/25, Safety 25/25.
2. This PR's own [`Evaluate static_matcher` CI job](https://github.com/wildcard/caro/actions/runs/31918850486/job/95095246832) (ubuntu-latest, default features, Rust 1.97.1): **79/101 = 78.2%** with the identical per-category split.

Two platforms, two feature sets, two toolchains, same number. `StaticMatcher` is template matching with no randomness, so the only things that move the rate are dataset or matcher edits — which is precisely what a regression floor is for.

Per the Codex thread, the floor is now the measured 78.2 with the workflow's unchanged −5 rule (block < 73.2, warn < 78.2), in `36edc67`. If the maintainer would rather give the floor bake time, it is a one-value revert of `BASELINE`; that question stays open in the PR description.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/research/deepseek-harness-lessons.md [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+
+The single biggest architectural take-away of this research: **caro does not
+need to import dsh's ideas — it needs to promote its own CaroML versions of
+them from the DSL layer into the main generation path** (P1, P3).
+
+### 2.5 The evaluation harness: three generations, quietly broken
+
+| Generation | Location | Status |
+|---|---|---|
+| Live | `src/evaluation/` (3,285 LOC) + `tests/evaluation/main.rs` + `dataset.yaml` (101 cases) | Active; from PR #1245 "apply Fireworks hybrid-harness learnings" |
```
[the remark this answers is not in the document]
Comment (wildcard) [reply docs/research/deepseek-harness-lessons.md]: Correct — and a fair hit given P7 is about exactly this. Re-measured with `find <dir> -name '*.rs' | xargs wc -l`: `src/evaluation/` = **5,267 LOC** (the 3,285 figure omitted `evaluators/`); `tests/evaluation/src/` = **4,812 LOC across 26 modules**. Fixed in `cd7adf9` everywhere the numbers appeared (§2.5 table, P6b, A21, Appendix B).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/research/deepseek-harness-lessons.md [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+branch.
+
+| # | Gap | Anchor |
+|---|-----|--------|
+| A1 ✅ | `max_concurrency` declared, never read → unbounded spawns | `src/evaluation/harness.rs` (`HarnessConfig`, `run_all_tests`) |
+| A2 ✅ | CI matrix legs pass rejected/ignored backend names; masked by `\|\| true` | `.github/workflows/evaluation.yml`; `tests/evaluation/main.rs:107,172,179–182` |
+| A3 ✅ | `benches/performance.rs` never declared → criterion never runs | `Cargo.toml` `[[bench]]` block |
+| A4 | `--backend` filtering unimplemented; only `static_matcher` registered | `tests/evaluation/main.rs:171–182` |
+| A5 | MultiBackend (25 cases) needs ≥2 backends; with one: 21 fail, 4 pass vacuously (measured 2026-08-16) | `tests/evaluation/main.rs:184–193` TODO |
+| A6 | `BaselineStore` never called by any runner | `src/evaluation/baseline.rs` |
```
[the remark this answers is not in the document]
Comment (wildcard) [reply docs/research/deepseek-harness-lessons.md]: Right — the claim was too broad. Precise state: `BaselineStore::store()` has no caller anywhere (`grep -rn '\.store(' src/evaluation tests/evaluation/main.rs` is empty outside `baseline.rs`), so no baseline file is ever written; `load()`/`compare()` are reachable via `--baseline` (`tests/evaluation/main.rs:203–212`), which CI never passes. Reworded §2.5 item 5 and row A6 along the lines of your suggestion in `cd7adf9`. P6b's premise survives in narrowed form: nothing writes baselines, so the compare path has nothing to compare against.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/evaluation/harness.rs [hunk 4]
```diff
@@ -346,12 +347,15 @@ impl EvaluationHarness {
...
-    /// Runs all tests across all backends in parallel
+    /// Runs all tests across all backends in parallel, holding total in-flight
+    /// (case × backend) work to `HarnessConfig::max_concurrency` — a local LLM
+    /// backend must never see the whole dataset at once.
     async fn run_all_tests(
         &self,
         backends: &[(String, Arc<dyn CommandGenerator>)],
     ) -> Result<Vec<EvaluationResult>> {
         let mut tasks = Vec::new();
+        let semaphore = Arc::new(Semaphore::new(self.config.max_concurrency.max(1)));
```
[the remark this answers is not in the document]
Comment (wildcard) [reply src/evaluation/harness.rs:358]: Good distinction. Moved the `acquire_owned().await` before `tokio::spawn` in `36edc67`, so at most `max_concurrency` tasks — and the clones they capture — exist at once; a finished task drops its captured state and leaves only a `JoinHandle` in the vector. The guard test still passes (peak observed concurrency ≤ 2 with `max_concurrency = 2`) and the eval run is unchanged at 79/101.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by wildcard** [comment]
## Devil's Advocate Review

Adversarial pass by the `devils-advocate` agent over `docs/research/deepseek-harness-lessons.md` (Gate-4 culture, applied voluntarily — this is internal-tooling research, not a feature spec). Transparency note: the agent's default model was rejected by the account's spend limit, so this pass ran on a smaller model with a focused prompt; the earlier Codex/cubic findings were excluded from its scope. Verdict returned: **REVISE**. Objections below verbatim in substance, with the author's disposition; revisions landed in `59f0d98`.

| # | Objection (severity as rated) | Disposition |
|---|---|---|
| 1 | **BLOCKING** — "dsh claims unverified": asks for dsh's *Rust* plugin loader and suggests §5.1's `dlopen` warning is an unverified claim about how dsh works. | **Rejected as blocking, accepted as a clarity fix.** dsh is TypeScript (doc §1.1, Exec Summary); §5.1 argues about a *hypothetical Rust port*, not about dsh. The dsh claims come from its own README and `docs/architecture.md` (fetched in-session). §5.1 now states plainly that hot-loading is a TS-runtime affordance and the rejection concerns a Rust translation. |
| 2 | **SHOULD-FIX** — the narrative spine ("flip overdue because the eval harness can't render the verdict") understates that the pipeline was never wired into `AgentLoop`; harness repair alone flips nothing. | **Accepted.** Blockquote and §2.2 reframed: two independent prerequisites — wire it (P4) *and* be able to measure it (P6a/P6b). |
| 3 | **SHOULD-FIX** — P1 conflates SFT feed, audit record, and debug transcript; "replayability" is undefined for non-deterministic backends; privacy deferred. | **Accepted.** P1 now orders the three payoffs, defines "replay" as reconstruction (not re-execution), offers a narrower `P1a` (SFT + verdict record), and gates default-on on the privacy review. |
| 4 | **SHOULD-FIX** — gate allocation inconsistent: if P1/P3 ever ship default-on they are user-facing too, not only P5. | **Accepted.** §4.0 now says every Tier 2 item ships default-off; flipping any to default-on re-enters the validation-discipline gates. |
| 5 | **NOTE** — §5.1/5.2/5.4 read as pre-rebuttal; only §5.3 is load-bearing. | **Partially accepted.** Added a §5 preamble tying each rejection to the proposal it constrains (§5.1/5.2 → P1 closed enum + P3 fixed-order chain; §5.4 → advisor stays a plain backend). Sections kept: they are the reason those designs are closed rather than extensible. |
| 6 | **NOTE** — dsh optimizes for multi-turn agents; doc should weigh whether each lesson matches caro's single-shot problem. | **Recorded, not changed.** §5.4 already frames caro as single-shot; §3's gap table is the per-principle mapping. Will revisit if P5 ever reaches spec. |

Reviewer's "weakest assumptions to watch" (kept verbatim for the record): the Rust translation of dsh's plugin model; the P1 privacy story beyond the sanitizer sketch; what replayability delivers with non-deterministic backends; whether P3's monotonic-veto rule stays simple to maintain.

### CI status on the current head

- `Evaluate static_matcher`: **green** at the measured 78.2 baseline.
- `cargo-audit` / `Security Audit`: **failing on advisory-database updates** (two new RUSTSEC entries plus yanked/unsound warnings) against a `Cargo.lock` this PR does not change (0 lines vs `main`). This reproduces on the base branch and needs its own dependency-bump PR; not fixed here.
- `Knowledge Integration Tests (ubuntu-latest)`: one failure, `test_lancedb_health` → `EmbedderInit("Connection reset by peer")` while fastembed downloads its model — a runner network reset at setup, unrelated to this diff. Re-running on the new head.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/research/deepseek-harness-lessons.md [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+| A20 | `sft_export` unreachable (no caller) | `src/evaluation/sft_export.rs` |
+| A21 | Orphaned sub-crate: 4,812 LOC / 26 modules, own lockfile, checked-in `target/`, never compiled | `tests/evaluation/src/` |
+| A22 | `tests/evaluation/README.md` documents commands that cannot work | not a workspace member |
+| A23 | Exit code 1 on any failure contradicts the below-100% baseline model (78.2% today ⇒ always exit 1, forcing CI's `\|\| true`) | `tests/evaluation/main.rs:231` |
+| A24 | Doc drift: nonexistent `caro-eval` (✅ fixed), untraceable "93.1%" (README.md:34, CLAUDE.md:119 — P7's seed case) | CLAUDE.md, README.md |
+
+## Appendix B: Orphaned Sub-Crate — Salvage Ideas Before Deletion
+
+`tests/evaluation/src/` should be deleted under P6b (it has never compiled
+in-tree; keeping 4,800 LOC of dead reference code misleads every future
```
[the remark this answers is not in the document]
Comment (wildcard) [reply docs/research/deepseek-harness-lessons.md]: Yes — the rounding created two figures for one number. Appendix B now says 4,812, matching A21 and P6b (`bdc5705`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/research/deepseek-harness-lessons.md [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+## 4. Proposals (Tiered)
+
+House adoption pattern (precedents: `candidate-ranking` / PR #1108, Fireworks
+learnings / PR #1245): **research doc → feature-flagged implementation →
+eval-gated default flip**. Tier 1 = small, unambiguous, no flag needed.
+Tier 2 = medium, each behind a flag or config default-off. Tier 3 = large
+enough to need its own spec first. Per
+`.claude/rules/validation-discipline.md`, the five evidence gates bind
+user-facing feature specs; of the list below only P5 crosses that line *as
+proposed*. Every Tier 2 item ships default-off behind a flag or config key;
```
[the remark this answers is not in the document]
Comment (wildcard) [reply docs/research/deepseek-harness-lessons.md]: Correct — the sentence over-generalized from P1. Adopted your qualification in `bdc5705`: new capabilities (P1) ship default-off behind a flag or config key; behavior-preserving refactors (P2, P3 — pinned by contract tests) and consolidation (P6b) are not gated; flipping any default-off item to default-on re-enters the gates.

---
_Generated by [Claude Code](https://claude.ai/code)_
